Feature: Make `ListView` the canonical encoding for `List`

connortsui20 opened the pull request:
Tracking Issue: https://github.com/vortex-data/vortex/issues/4699

Revived from: https://github.com/vortex-data/vortex/pull/4853 and https://github.com/vortex-data/vortex/pull/4830

This PR makes `ListViewArray` the canonical encoding of `DType::List`, which used to be `ListArray`.

We will probably have a large regression in term of performance for now, but that can be improved in the future.

connortsui20 committed:
add `Clone` to all scalars

Signed-off-by: Connor Tsui <connor.tsui20@gmail.com>

connortsui20 committed:
make listview canonical over list (just in vortex-array)

Signed-off-by: Connor Tsui <connor.tsui20@gmail.com>

connortsui20 committed:
make listview canonical over list (in remaining crates)

Signed-off-by: Connor Tsui <connor.tsui20@gmail.com>

connortsui20 committed:
clean up `listview` implementation

Signed-off-by: Connor Tsui <connor.tsui20@gmail.com>

connortsui20 committed:
clean up listview tests

Signed-off-by: Connor Tsui <connor.tsui20@gmail.com>

connortsui20 commented on vortex-array/src/arrays/listview/array.rs:
```diff
@@ -345,74 +352,3 @@ where
     Ok(())
 }
-
-/// Create a [`ListViewArray`] from a [`ListArray`](crate::arrays::ListArray) by computing `sizes`
-/// from `offsets`.
-pub fn list_view_from_list(list: ListArray) -> ListViewArray {
```
Moved to `conversion.rs`

connortsui20 commented on vortex-array/src/compute/list_contains.rs:
I want to clean this file up in a separate PR

a10y commented on vortex-array/src/arrays/chunked/vtable/canonical.rs:
```diff
@@ -82,56 +82,68 @@ fn swizzle_struct_chunks(
+    // TODO(connor)[ListView]: We could potentially choose a smaller type here, but that would make
+    // this much more complicated.
+
+    // We (somewhat arbitrarily) choose `u64` for our offsets and sizes here.
```
this seems fine to me, we can always .narrow() at the end

joseph-isaacs replied:
You can find the max size and offset DType and use that? I think that might be worth it to hopefully avoid a  few casts.

connortsui20 replied:
Dan has a macro in `encodings/sparse/src/canonical.rs` called `match_smallest_offset_type` that I was thinking about extracting it out into `vortex-dtype` so we can use it everywhere. I left a TODO in there and I can look at it later.

a10y commented on vortex-array/src/arrays/list/compute/take.rs:
```diff
@@ -1,179 +1,30 @@
-/// that lists are stored contiguously and in-order (`offset[i+1] >= offset[i]`). Taking
-/// non-contiguous indices would violate this requirement.
+/// This implementation converts the [`ListArray`] to a [`ListViewArray`] and then delegates to its
+/// `take` implementation. This approach avoids the need to rebuild the `elements` array.
```
for very short indices arrays, it'd probably be more efficient to build the offsets/lengths incrementally. but this is probably fine for now

connortsui20 replied:
I left a TODO in there

a10y commented on vortex-array/src/arrays/listview/compute/take.rs:
```diff
@@ -72,10 +74,11 @@ impl TakeKernel for ListViewVTable {
 
-        Ok(new_array.into_array())
+        Ok(new_array
+            .rebuild(ListViewRebuildMode::MakeZeroCopyToList)
```
why are we doing this here? we should lazily compact the ListView not every time we do a take. there should be something like the `compact` method on VarBinViewArray. This is something we want todo at compression time but not during query processing probably?

connortsui20 replied:
See the response in the other conversation, but I also added a TODO comment that was more descriptive

a10y commented on vortex-array/src/canonical.rs:
```diff
@@ -78,14 +78,14 @@ pub enum Canonical {
     VarBinView(VarBinViewArray),
-    // TODO(connor)[ListView]: Convert the canonical encoding of `List` to `ListViewArray`.
-    List(ListArray),
+    List(ListViewArray),
```
perhaps we should change the variant name here to ListView, similar to VarBinView

joseph-isaacs replied:
I think these would match where possible the DType?

connortsui20 replied:
We talked about it offline with Nick, these should have the same name as `DType`, with the exception of binary data. Arguably, it shouldn't even be called `VarBinView` and instead be called something like `Canonical::Binary` or `Canonical::String`? Regardless, I added some more documentation at the top to make it clearer.

a10y replied:
that's fine, i just like consistency

a10y commented on vortex-array/src/context.rs:
```diff
@@ -25,21 +25,21 @@ impl ArrayRegistry {
             EncodingRef::new_ref(DecimalEncoding.as_ref()),
             EncodingRef::new_ref(StructEncoding.as_ref()),
-            EncodingRef::new_ref(ListEncoding.as_ref()),
+            EncodingRef::new_ref(ListViewEncoding.as_ref()),
```
The comment is sadly wrong, this should contain all of the builtin encodings from the vortex-array crate. See below how we include VarBinEncoding. I think similarly, we should include ListEncoding here as well

connortsui20 replied:
I moved some things around so we do canonical first then everything else below. It seems like we are missing `DatetimeEncoding` too?

connortsui20 replied:
Oh datetime is an extension type, nevermind

a10y commented on vortex-dtype/src/arrow.rs:
```diff
@@ -127,7 +127,10 @@ impl FromArrowType<(&DataType, Nullability)> for DType {
+            DataType::List(e)
+            | DataType::LargeList(e)
+            | DataType::ListView(e)
+            | DataType::LargeListView(e) => {
```
below in to_arrow_dtype needs to be updated to return DataType::ListView

connortsui20 replied:
nice catch!

a10y commented on vortex-scalar/src/binary.rs:
```diff
@@ -15,7 +15,7 @@ use crate::{InnerScalarValue, Scalar, ScalarValue};
 /// This type provides a view into a binary scalar value, which can be either
 /// a valid byte buffer or null.
-#[derive(Debug, Hash)]
+#[derive(Debug, Clone, Hash)]
```
why did you make all of these clone?

connortsui20 replied:
https://github.com/vortex-data/vortex/issues/4844

I think I needed clone on list scalars for something in this change, and so I just added it to everything.

a10y commented on vortex-array/src/arrays/listview/compute/filter.rs:
```diff
@@ -57,10 +59,11 @@ impl FilterKernel for ListViewVTable {
 
-        Ok(new_array.into_array())
+        Ok(new_array
+            .rebuild(ListViewRebuildMode::MakeZeroCopyToList)
```
same deal here: i don't think we should try and optimize/gc the elements when we run filter(ListView)

joseph-isaacs replied:
I think we should do this, otherwise when we later export we end up with a load of unused data we need to operate on (when we canoncalize)

connortsui20 replied:
The reason I wrote it like this is because if we don't do this, we get an even larger regression whenever we try and export data to something (like DuckDB) because, until we have operators, we cannot decide a "good" point where we want to garbage collect / rebuild. Obviously a good point is just after we have run / pipelined a bunch of filters and takes in succession, but we can't really do that with the current architecture.

For now I think we should just keep the TODO in there and we can revisit this.

joseph-isaacs commented on vortex-array/src/arrays/chunked/vtable/canonical.rs:
```diff
@@ -82,56 +82,68 @@ fn swizzle_struct_chunks(
+        let current_elements_offset = elements_builder.len() as u64;
+
+        // Append all elements from this chunk.
+        elements_builder.extend_from_array(chunk.elements());
```
I think we want to use a chunked array of the elements and update the offsets

connortsui20 replied:
Added this in https://github.com/vortex-data/vortex/pull/4946/commits/c18c40ec29e10bf4122d382c4da94cea649dcb14

joseph-isaacs commented on vortex-array/src/arrays/constant/vtable/canonical.rs:
```diff
@@ -204,58 +208,54 @@ fn canonical_byte_view(scalar_bytes: Option<&[u8]>, dtype: &DType, len: usize) -
+    for _ in 0..len {
+        listview_builder
+            .append_scalar(scalar)
+            .vortex_expect("somehow unable to append list scalar to listview builder");
```
Here we can return a LIstView with a constant size, element and offset arrays.

connortsui20 replied:
https://github.com/vortex-data/vortex/pull/4946/commits/1b7a179f68a76177114c796a94859fdc80e7f225

joseph-isaacs commented on vortex-btrblocks/src/lib.rs:
```diff
@@ -399,18 +399,29 @@ impl BtrBlocksCompressor {
                 let compressed_offsets = IntCompressor::compress_no_dict(
                     &list_array.offsets().to_primitive().narrow()?,
                     false,
                     MAX_CASCADE,
```
Do we want `MAX_CASCADE` to reset for structural types?

connortsui20 replied:
Not sure what the alternative should be, `MAX_CASCADE - 1`?

joseph-isaacs commented on vortex-duckdb/src/exporter/list.rs:
```diff
@@ -15,74 +16,125 @@ use crate::cpp;
+            let elements_exporter = new_array_exporter_with_flatten(array.elements(), cache, true)?;
+
+            // TODO(connor)[4809]: THIS IS A HACK, we should be checking this everywhere.
+            if !elements.is_empty() {
+                elements_exporter.export(0, elements.len(), &mut duckdb_elements)?;
+            }
```
can we just fix this, since I am not sure this exporter will handle empty size/offsets

connortsui20 replied:
I will leave it to this: https://github.com/vortex-data/vortex/issues/4809

If it was just list I would fix this, but because the dict exporter is broken in the exact same way (I forgot that I just `ignore` the tests for both list and dict) I'd rather fix them together.

joseph-isaacs commented on vortex-layout/src/layouts/compact.rs:
```diff
@@ -132,19 +133,19 @@ impl CompactCompressor {
-                // recurse
                 let compressed_elems = self.compress(list_array.elements())?;
                 let compressed_offsets = self.compress(list_array.offsets())?;
+                let compressed_sizes = self.compress(list_array.sizes())?;
```
Does this narrow?

connortsui20 replied:
I'm actually now questioning if we should be calling `narrow()` in `vortex-btrblocks`. Isn't it somewhat unexpected that we write one type and then get back a different type? I guess it's probably fine because the types of the offsets and sizes are not encoded in our `DType`. But yes both should be the same.

connortsui20 replied:
Added narrowing in https://github.com/vortex-data/vortex/pull/4946/commits/f34587bc4e1f686cd2210763162032b4ab20113a

connortsui20 committed:
address some comments

Signed-off-by: Connor Tsui <connor.tsui20@gmail.com>

connortsui20 committed:
make canonicalize of constant list view array more efficient

Signed-off-by: Connor Tsui <connor.tsui20@gmail.com>

connortsui20 committed:
make canonicalize of chunked list view use chunked elements

Signed-off-by: Connor Tsui <connor.tsui20@gmail.com>

connortsui20 committed:
narrow offsets and sizes in layouts compressor

Signed-off-by: Connor Tsui <connor.tsui20@gmail.com>

connortsui20 committed:
clean up

Signed-off-by: Connor Tsui <connor.tsui20@gmail.com>

connortsui20 commented:
So... we hit this in the parquet implementation 😭:

https://github.com/apache/arrow-rs/blob/751b0822a7f0b2647c1c662131131b35f268bfef/parquet/src/arrow/schema/mod.rs#L682-L684

connortsui20 committed:
convert list view to list in compression benchmark for parquet

Parquet rust doesn't seem to support list view, so we convert to list
right before we write to Parquet.

Signed-off-by: Connor Tsui <connor.tsui20@gmail.com>

connortsui20 commented:
Now ready for re-review, I'm pretty sure I hit everything, and everything I didn't hit is either in a TODO or an issue

robert3005 commented:
This parquet todo is fabulous

connortsui20 committed:
touch ups / style clean up

Signed-off-by: Connor Tsui <connor.tsui20@gmail.com>

connortsui20 committed:
use `UninitRange` for list builder extend

Signed-off-by: Connor Tsui <connor.tsui20@gmail.com>

connortsui20 committed:
rename `set_bit` to `set_validity_bit`

Signed-off-by: Connor Tsui <connor.tsui20@gmail.com>

robert3005 reviewed (approved):
Hard to review such a big change but multiple people left comments throughout and I think at this point it's best to merge this since we need listview as canonical list and work on whatever we find in followups
